go/vault/api/vault.go

```diff
@@ -0,0 +1,135 @@
+	CreatedNonce uint64 `json:"created_nonce"`
+	// State is the vault state.
+	State State `json:"state"`
+	// Nonce is the nonce to use for the next action.
+	Nonce uint64 `json:"nonce,omitempty"`
+
+	// AdminAuthority specifies the vault's admin authority.
+	AdminAuthority Authority `json:"admin_authority"`
+	// SuspendAuthority specifies the vault's suspend authority.
+	SuspendAuthority Authority `json:"suspend_authority"`
+}
+
+// NewVaultAddress returns the address for the vault.
+func NewVaultAddress(createdBy staking.Address, createdNonce uint64) staking.Address {
+	return staking.NewModuleAddress(ModuleName, fmt.Sprintf("vault.%s.%d", createdBy, createdNonce))
+}
+
+// Address returns the address for the vault.
+func (v *Vault) Address() staking.Address {
+	return NewVaultAddress(v.CreatedBy, v.CreatedNonce)
+}
+
+// IsActive retruns true iff the vault is currently active (processing withdrawals).
+func (v *Vault) IsActive() bool {
+	return v.State == StateActive
+}
+
+// Authorities returns the list of all vault authorities.
+func (v *Vault) Authorities() []*Authority {
+	return []*Authority{
+		&v.AdminAuthority,
+		&v.SuspendAuthority,
+	}
+}
+
+// Authority is the vault multisig authority.
+type Authority struct {
+	// Addresses are the addresses that can authroize an action.
+	Addresses []staking.Address `json:"addresses"`
+	// Threshold is the minimum number of addresses that must authorize an action.
+	Threshold uint8 `json:"threshold"`
+}
+
+// Validate validates the authority configuration.
+func (a *Authority) Validate(params *ConsensusParameters) error {
+	if len(a.Addresses) == 0 {
+		return fmt.Errorf("no addresses")
+	}
+	if a.Threshold == 0 {
+		return fmt.Errorf("threshold cannot be zero")
+	}
+	if int(a.Threshold) > len(a.Addresses) {
+		return fmt.Errorf("threshold is larger than the number of addresses")
+	}
+	if len(a.Addresses) > int(params.MaxAuthorityAddresses) {
+		return fmt.Errorf("too many addresses in authority (max: %d got: %d)",
+			params.MaxAuthorityAddresses, len(a.Addresses))
+	}
+
+	// Ensure no duplicate addresses.
+	addressMap := make(map[staking.Address]struct{})
+	for _, addr := range a.Addresses {
+		if _, ok := addressMap[addr]; ok {
+			return fmt.Errorf("duplicate address in authority: %s", addr)
+		}
+		addressMap[addr] = struct{}{}
+	}
+	return nil
+}
+
+// Contains checks whether the set of authorities contains the given address.
+func (a *Authority) Contains(address staking.Address) bool {
+	for _, addr := range a.Addresses {
+		if addr.Equal(address) {
+			return true
+		}
+	}
+	return false
+}
+
+// Verify checks whether the passed set of addresses is sufficient to authorize an action.
+func (a *Authority) Verify(addresses []staking.Address) bool {
+	addressMap := make(map[staking.Address]struct{})
+	for _, addr := range a.Addresses {
+		addressMap[addr] = struct{}{}
+	}
+	var count int
+	for _, addr := range addresses {
```

Comment:
This list needs to be deduplicated (or remove the address from `addressMap` on first match).

Reply:
Also add a test for this case.

Reply:
Duplicate addresses are rejected on validation, there is already a test case for this, but let me check it may only be in transaction tests.

Reply:
But agree that it is better to do this in both places so it works correctly even in case validation was not performed.

Reply:
Oh I see, you mean duplicate addresses in `addresses`, yeah let me fix that.